modules/apps/transfer/keeper/msg_server.go

```diff
@@ -59,3 +66,33 @@ func (k Keeper) UpdateParams(goCtx context.Context, msg *types.MsgUpdateParams)
 
 	return &types.MsgUpdateParamsResponse{}, nil
 }
+
+// unwindHops unwinds the hops present in the tokens denomination and returns the message modified to reflect
+// the unwound path to take. It assumes that only a single token is present (as this is verified in ValidateBasic)
+// in the tokens list and ensures that the token is not native to the chain.
+func (k Keeper) unwindHops(ctx sdk.Context, msg *types.MsgTransfer) (*types.MsgTransfer, error) {
+	coins := msg.GetCoins()
+	token, err := k.tokenFromCoin(ctx, coins[0])
+	if err != nil {
+		return nil, err
+	}
+
+	if token.Denom.IsNative() {
+		return nil, errorsmod.Wrap(types.ErrInvalidForwarding, "cannot unwind a native token")
+	}
+	var unwindHops []types.Hop
+	// remove the first hop in denom as it is the current port/channel on this chain
+	for _, trace := range token.Denom.Trace[1:] {
+		unwindHops = append(unwindHops, types.Hop{PortId: trace.PortId, ChannelId: trace.ChannelId}) //nolint: gosimple
+	}
+
+	// Update message fields.
+	msg.SourcePort, msg.SourceChannel = token.Denom.Trace[0].PortId, token.Denom.Trace[0].ChannelId
+	msg.Forwarding.Hops = append(unwindHops, msg.Forwarding.Hops...)
+	msg.Forwarding.Unwind = false
+
+	if err := msg.ValidateBasic(); err != nil {
+		return nil, err
+	}
```

Comment:
[nit] maybe we should add a comment that the only real reason this could fail, is if exceeding 16 hops by using the unwind field.